src/hooks/init/check-update.ts

```diff
@@ -1,69 +1,135 @@
 import {Hook} from '@oclif/core'
-import type {Chalk} from 'chalk'
-import {spawn} from 'child_process'
-import {readFile, stat} from 'fs/promises'
-import {join} from 'path'
+import chalk from 'chalk'
+import makeDebug from 'debug'
+import {spawn} from 'node:child_process'
+import {Stats} from 'node:fs'
+import {readFile, stat, utimes} from 'node:fs/promises'
+import {dirname, join, resolve} from 'node:path'
+import {fileURLToPath} from 'node:url'
+
+const readJSON = async <T>(file: string): Promise<{contents: T; stat: Stats}> => ({
+  contents: JSON.parse(await readFile(file, 'utf8')),
+  stat: await stat(file),
+})
+
+function hasNotBeenMsSinceDate(ms: number, date: Date): boolean {
+  const now = new Date()
+  const diff = now.getTime() - date.getTime()
+  return diff < ms
+}
+
+function convertToMs(frequency: number, unit: 'days' | 'hours' | 'milliseconds' | 'minutes' | 'seconds'): number {
+  switch (unit) {
+    case 'days': {
+      return frequency * 24 * 60 * 60 * 1000
+    }
+
+    case 'hours': {
+      return frequency * 60 * 60 * 1000
+    }
+
+    case 'milliseconds': {
+      return frequency
+    }
+
+    case 'minutes': {
+      return frequency * 60 * 1000
+    }
+
+    case 'seconds': {
+      return frequency * 1000
+    }
+
+    default: {
+      // default to minutes
+      return frequency * 60 * 1000
+    }
+  }
+}
 
 const hook: Hook<'init'> = async function ({config}) {
+  const debug = makeDebug('update-check')
   const file = join(config.cacheDir, 'version')
 
   // Destructure package.json configuration with defaults
   const {
-    timeoutInDays = 60,
+    authorization = '',
+    frequency,
+    frequencyUnit,
     message = '<%= config.name %> update available from <%= chalk.greenBright(config.version) %> to <%= chalk.greenBright(latest) %>.',
     registry = 'https://registry.npmjs.org',
-    authorization = '',
-  } = (config.pjson.oclif as any)['warn-if-update-available'] || {}
+    timeoutInDays = 60,
+  } = config.pjson.oclif['warn-if-update-available'] || {}
+
+  const warningFrequency = (config.scopedEnvVar('NEW_VERSION_CHECK_FREQ') as number | undefined) ?? frequency
```

Comment:
TS lets you make ASsertions about what's in somebody's env (lying to the compiler), but that doesn't do any checks or conversions in the real world.

example: if their frequency was `ham` and the unit was `sandwiches` then `convertToMs` is going to have unhandled errors.

Also, envs would be a string anyway, so you'll need to parseInt `warningFrequency`, right?

Reply:
let's have some tests for handling of these envs.